src/Language/DifferentialDatalog/Syntax.hs

```diff
@@ -289,21 +323,37 @@ instance WithName TypeDef where
 
 instance PP TypeDef where
     pp TypeDef{..} | isJust tdefType
-                   = "typedef" <+> pp tdefName <>
-                     (if null tdefArgs
-                         then empty
-                         else "<" <> (hcat $ punctuate comma $ map (("'" <>) . pp) tdefArgs) <> ">") <+>
-                     "=" <+> (pp $ fromJust tdefType)
-    pp TypeDef{..} = "extern type" <+> pp tdefName <>
-                     (if null tdefArgs
-                         then empty
-                         else "<" <> (hcat $ punctuate comma $ map (("'" <>) . pp) tdefArgs) <> ">")
+                   = ppAttributes tdefAttrs $$
+                     ("typedef" <+> pp tdefName <>
+                      (if null tdefArgs
+                          then empty
+                          else "<" <> (hcat $ punctuate comma $ map (("'" <>) . pp) tdefArgs) <> ">") <+>
+                      "=" <+> (pp $ fromJust tdefType))
+    pp TypeDef{..} = ppAttributes tdefAttrs $$   
+                     ("extern type" <+> pp tdefName <>
+                      (if null tdefArgs
+                          then empty
+                          else "<" <> (hcat $ punctuate comma $ map (("'" <>) . pp) tdefArgs) <> ">"))
 
 instance Show TypeDef where
     show = render . pp
 
 instance Eq TypeDef where
-    (==) t1 t2 = name t1 == name t2 && tdefType t1 == tdefType t2
+    (==) t1 t2 = (name t1, tdefAttrs t1, tdefAttrs t1, tdefType t1) == (name t2, tdefAttrs t2, tdefAttrs t2, tdefType t2)
+
+tdefCheckSizeAttr :: (MonadError String me) => TypeDef -> me (Maybe Int)
+tdefCheckSizeAttr TypeDef{..} =
```

Comment:
if i understand this right the size attribute will be ignored on types which are not extern

Reply:
moreover, it won't be checked.

Reply:
`Validate.hs` has code to check that only extern types can have this attribute.  This function only checks that the attribute has correct format.